attempt 1, failing with no SO_NOSIGPIPE found and implicit declaratio…

jpbland1 opened the pull request:
…n of pread

billphipps commented on src/wh_server.c:
```diff
@@ -4,18 +4,20 @@
-#include "wolfssl/wolfcrypt/rsa.h"
-#include "wolfssl/wolfcrypt/aes.h"
-#include "wolfssl/wolfcrypt/hmac.h"
+    #include <wolfssl/options.h>
 #endif
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/wolfcrypt/types.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/wc_port.h>
+#include <wolfssl/wolfcrypt/sha256.h>
+#include <wolfssl/wolfcrypt/aes.h>
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/cmac.h>
+#include <wolfssl/wolfcrypt/kdf.h>
+#include <wolfssl/wolfcrypt/cryptocb.h>
```
Let's leave "" here

jpbland1 replied:
done

billphipps commented on src/wh_server.c:
```diff
@@ -28,8 +30,573 @@
+#include "wolfhsm/wh_packet.h"
 #include "wolfhsm/wh_server.h"
 
+#ifdef WOLFHSM_SHE_EXTENSION
```
Recommend to move all of the SHE stuff to a separate file.  Ensure the any state variables are members of the server context to avoid globals.

billphipps commented on src/wh_server.c:
```diff
@@ -28,8 +30,573 @@
+Cmac sheCmac[1];
+#endif
+
+#if 0
```
Can you can leave out any dormant code for now since we won't be reviewing it.

jpbland1 replied:
no, want to get this all in the repo ASAP, it will come out as we get it working

billphipps commented on src/wh_server.c:
```diff
@@ -54,7 +621,7 @@ int wh_Server_Init(whServerContext* server, whServerConfig* config)
 static int _wh_Server_HandleCommRequest(whServerContext* server,
         uint16_t magic, uint16_t action, uint16_t seq,
         uint16_t req_size, const void* req_packet,
-        uint16_t *out_resp_size, void* resp_packet)
```
As a style point, I'll normally leave the * attached to an "output" parameter.  Ok to change if this is confusing.

jpbland1 replied:
`(void*)myVal` or `(void *)myVal`? makes no sense putting part of the type on the variable name

billphipps commented on src/wh_server.c:
```diff
@@ -119,35 +686,1428 @@ static int _wh_Server_HandleNvmRequest(whServerContext* server,
     return rc;
 }
 
+static int _wh_Server_HandleCryptoRequest(whServerContext* server,
```
Consider moving this crypto handler to a separate file to simplify management.

billphipps commented on src/wh_server.c:
```diff
@@ -119,35 +686,1428 @@ static int _wh_Server_HandleNvmRequest(whServerContext* server,
-        uint16_t magic, uint16_t action, uint16_t seq,
-        uint16_t req_size, const void* req_packet,
-        uint16_t *out_resp_size, void* resp_packet)
+#ifdef WOLFHSM_SHE_EXTENSION
```
Move to a separate SHE source file

billphipps reviewed (changes requested):
The SO_NOSIGPIPE was removed previously and replaced using send/recv with MSG_NOSIGNAL flags in the Posix TCP transport. You'll probably need to pull and rebase to correct that. pread/pwrite are supposed to be in unistd.h . Make sure the tests build and run cleanly in your build environment before adding stuff.

Please move all of the crypto and SHE functions into separate files wh_crypto.h/.c and wh_she.h/.c .  Please move message definitions/translations into wh_message_crypto.h/.c and wh_message_she.h/.c

bigbrett reviewed (changes requested):
@jpbland1 Looks like you added this to an old version of the code, could you rebase onto main? I think a lot of your errors will go away. Tests have been split out too.

billphipps commented on wolfhsm/wh_cryptocb.h:
```diff
@@ -0,0 +1,35 @@
+ */
+#ifndef WOLFHSM_CRYPTOCB_H
+#define WOLFHSM_CRYPTOCB_H
+#include <wolfhsm/common.h>
```
```suggestion
#include "wolfhsm/common.h"
```

Also, make sure to move all of the necessary includes for this file into here.

jpbland1 replied:
done

billphipps commented on wolfhsm/wh_cryptocb.h:
```diff
@@ -0,0 +1,35 @@
+#define WOLFHSM_CRYPTOCB_H
+#include <wolfhsm/common.h>
+
+#ifdef __cplusplus
+    extern "C" {
+#endif
```
Let's remove __cplusplus guards for now.  We use C99 initializers which are not supported by c++.

billphipps commented on wolfhsm/wh_cryptocb.h:
```diff
@@ -0,0 +1,35 @@
+ *
+ * wolfSSL is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
```
All of wolfHSM is actually under GPLv3, not 2.  We'll go back and fix all of the licensing once we are mostly feature complete.

billphipps commented on wolfhsm/wh_packet.h:
```diff
@@ -0,0 +1,456 @@
+ */
+#ifndef WOLFHSM_PACKET_H
+#define WOLFHSM_PACKET_H
+#include <wolfhsm/common.h>
```
Same header include comment

jpbland1 replied:
done

billphipps commented on wolfhsm/wh_packet.h:
```diff
@@ -0,0 +1,456 @@
+    uint32_t version;
+} wh_Packet_version_exchange;
+
+#ifdef WOLFHSM_SHE_EXTENSION
```
Let's move all of the SHE stuff to a separate .h/.c file

billphipps commented on wolfhsm/wh_packet.h:
Recommend to change name to wh_message_crypto.h

billphipps replied:
Added to todo list

billphipps commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,474 @@
+#ifndef WOLFSSL_USER_SETTINGS
+    #include <wolfssl/options.h>
+#endif
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/wolfcrypt/types.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/aes.h>
+#include <wolfssl/wolfcrypt/cmac.h>
+#include <wolfssl/wolfcrypt/cryptocb.h>
+#include <wolfhsm/wh_packet.h>
+#include <wolfhsm/common.h>
+#include <wolfhsm/wh_error.h>
+#include <wolfhsm/wh_client.h>
+#include <wolfhsm/wh_message.h>
```
Recommend to use "" instead of <>

billphipps commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,474 @@
+                ret = packet->error;
+            /* read out */
+            else
+                XMEMCPY(info->rng.out, out, packet->rngRes.sz);
```
Recommend to check rngRes.sz for sanity before memcpy'ing

billphipps replied:
Added to todo list

billphipps commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,474 @@
+            do {
+                ret = wh_Client_RecvResponse(ctx, &group, &action, &dataSz,
+                    rawPacket);
+                sleep(1);
```
This is not portable.  Let's replace this with XUSLEEP(1000) macro that the user can replace.  Default implementation simply returns

billphipps replied:
Added to todo list

billphipps commented on test/user_settings.h:
Grumble.  Not sure we should have a user_settings here.  Lemme think about it.

billphipps replied:
Added to todo list

billphipps commented on test/wh_test_crypto.c:
This should be changed to use the memtransport so we don't need networking for test.  Leave for now and I'll put together a client server test that you can directly reuse.

billphipps commented on wolfhsm/common.h:
This file should be removed with its contents going in other spots.  consider wh_common.h

billphipps reviewed (changes requested):
Lots of integration nits, but we're getting there!

jpbland1 commented on test/wh_test.c:
```diff
@@ -587,13 +586,20 @@ static void* _whClientTask(void *cf)
+    /* register the wolfHSM crypto callback */
+    if (wc_CryptoCb_RegisterDevice(myDevId, wolfHSM_CryptoCb, ctx) != 0)
+        return NULL;
+
```
done

jpbland1 commented on test/wh_test.c:
```diff
@@ -576,6 +574,7 @@ void wh_CommClientServer_TcpThreadTest(void)
 
 static void* _whClientTask(void *cf)
 {
+    int myDevId = 1;
```
fixed

jpbland1 commented on wolfhsm/wh_server.h:
```diff
@@ -3,6 +3,11 @@
+    #include <wolfssl/options.h>
+#endif
+#include <wolfssl/wolfcrypt/settings.h>
+
```
done

jpbland1 commented on wolfhsm/wh_server.h:
```diff
@@ -11,9 +16,9 @@
 /* Context structure to maintain the state of an HSM server */
 typedef struct whServerContext_t {
     whCommServer comm[1];
+    WC_RNG rng[1];
```
done

jpbland1 commented on src/wh_server.c:
```diff
@@ -4,18 +4,20 @@
 
-#if 0
-/* wolfCrypt */
+#include <arpa/inet.h>
```
yes for a SHE function it must use big endian

billphipps replied:
Added to todo list

billphipps commented on wolfhsm/wh_server.h:
```diff
@@ -2,18 +2,51 @@
+typedef struct CacheSlot {
+    uint8_t commited;
+    NvmMetaData meta[1];
+    uint8_t buffer[WOLFHSM_NVM_MAX_OBJECT_SIZE];
```
This is not going to work on small RAM devices.  Max object size is 64kB.  The intent would be to dynamically load whatever content is needed during the request and unload it afterwards.  The NVM Read function allows partial reads at an offset so only the pertinent part of the object data must be loaded in RAM at any given time.  Let's work on an alternate cache design that doesn't necessarily rely on fully preallocating RAM storage.  Ok for now, but this is a design issue.

billphipps replied:
Added to todo list

billphipps reviewed (approved):
Still have a number of organizational/design comments, but we need to get this code put in now so we can continue to make progress.

bigbrett commented on src/wh_server.c:
```diff
@@ -28,8 +30,613 @@
+    uint16_t id;
+    /* try every index, unless both cache and nvm are full one will work */
+    for (id = 1; id < WOLFHSM_NUM_RAMKEYS + 1; id++) {
+        /* check against chache keys */
```
```suggestion
        /* check against cache keys */
```

bigbrett commented on wolfhsm/wh_server.h:
```diff
@@ -2,18 +2,51 @@
+#ifdef WOLFHSM_SHE_EXTENSION
+    uint32_t count;
+#endif
+} NvmMetaData;
```
Why are we defining a server-specific meta data struct here? Isn't [whNvmMetaData ](https://github.com/wolfSSL/wolfHSM/blob/main/wolfhsm/wh_common.h#L76)already defined in wh_common.h? I don't think we want multiple clones of this floating around

bigbrett commented on wolfhsm/wh_server.h:
```diff
@@ -2,18 +2,51 @@
+    curve25519_key curve25519Private[1];
+    curve25519_key curve25519Public[1];
+    WC_RNG rng[1]; 
+} crypto_context;
```
nit: we should probably follow the naming convention for public structs
```suggestion
}  whCryptoContext;
```

billphipps replied:
added to todo list

bigbrett commented on wolfhsm/wh_server.h:
```diff
@@ -2,18 +2,51 @@
+    uint8_t commited;
+    NvmMetaData meta[1];
+    uint8_t buffer[WOLFHSM_NVM_MAX_OBJECT_SIZE];
+} CacheSlot;
```
nit: we should probably follow naming convention for public structs
```suggestion
} whCacheSlot;
```

bigbrett commented on wolfhsm/wh_server.h:
```diff
@@ -2,18 +2,51 @@
+    curve25519_key curve25519Private[1];
+    curve25519_key curve25519Public[1];
+    WC_RNG rng[1]; 
+} crypto_context;
```
just musing (no action needed) but I wonder if this makes sense to define in its own file (e.g. `whCryptoContext.h`) since it will eventually grow to hold the state of every crypto algorithm we want to support

bigbrett commented on src/wh_server.c:
```diff
@@ -28,8 +30,613 @@
+    int foundIndex = -1;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
```
This comment is a catch-all for the whole file: I think we should be using wolfHSM error codes, not wolfSSL error codes. I think we only want to use wolfCrypt error codes to process result of wolfCrypt  functions

billphipps replied:
wolfCrypt errors should be returned from cryptocb.c, but wolfHSM everywhere else.  Our goal was to have very few error codes to make autonomous handling simple.  Broke or not broke.

billphipps replied:
Added to todo list

bigbrett commented on src/wh_server.c:
```diff
@@ -28,8 +30,613 @@
+    /* write key if slot found */
+    XMEMCPY((uint8_t*)server->cache[foundIndex].buffer, in, meta->len);
+    XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+        sizeof(NvmMetaData));
```
probably want `sizeof(server->cache[foundIndex].meta)`, especially since in my other comment I point out that we want to change the type to `whNvmMetaData`

bigbrett commented on src/wh_server.c:
```diff
@@ -28,8 +30,613 @@
+    }
+#endif
+#endif
+    return ret;
```
is it possible to not get a hit on the key in the cache? It looks like this unconditionally returns success.

jpbland1 replied:
the dead code handled the miss case, updated since nvm isn't working

bigbrett commented on src/wh_server.c:
```diff
@@ -28,8 +30,613 @@
+    }
+    if (ret == 0)
+        ret = meta->id;
+    return ret;
```
I think we are mixing up return codes and the unique ID. It should probably return an error code and set the ID as an output argument. We need to be able to return an error code in case the wolfcrypt calls fail etc.

jpbland1 replied:
errors are negative, all keyIds are positive and are limited to uint_16 so there's no overlap

bigbrett commented on src/wh_server.c:
```diff
@@ -119,35 +728,1492 @@ static int _wh_Server_HandleNvmRequest(whServerContext* server,
+            }
+            else
+                ret = BAD_FUNC_ARG;
+            break;
```
on success this will return the key ID not a valid return code. We need to refactor hsmCacheKey to return a proper error code, or at the very least, fix all cases of mixed up return code/key IDs in this function

bigbrett commented on wolfhsm/common.h:
```diff
@@ -0,0 +1,226 @@
+
+/* common.h
```
We already have `wh_common.h` I don't think we should add an additional common file

jpbland1 replied:
fixed

bigbrett reviewed (changes requested):
Good start, but I think we need to clean up a lot before it can be merged in. I would really like to see all the dead code blocks removed. You can have a local branch that has them not removed, but we should be adding that code in piecemeal, rather than throwing it all in and commenting/ifdefing it out.

jpbland1 commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,531 @@
+            packet->pkCurve25519kgReq.sz = info->pk.curve25519kg.size;
+            /* write request */
+            ret = wh_Client_SendRequest(ctx, group,
+                WC_PK_TYPE_CURVE25519_KEYGEN, sizeof(packet->rngReq),
```
yes, fixed

jpbland1 commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,531 @@
+            } while (ret == WH_ERROR_NOTREADY);
+        }
+        if (ret == 0) {
+            if (group != WH_MESSAGE_GROUP_CRYPTO || action != WC_ALGO_TYPE_RNG)
+                ret = packet->error;
```
my mistake, fixed and set error

bigbrett commented on src/wh_cryptocb.c:
```diff
@@ -400,12 +82,15 @@ int wolfHSM_CryptoCb(int devId, wc_CryptoInfo* info, void* inCtx)
             }
```
Still need to remove the `sleep()`s from this function. Line 398 for example (sorry, GH won't let me comment on that line in this review since it wasn't changed)

bigbrett reviewed (dismissed):
Nice, some good fixes! Did a quick pass with some things that might have been overlooked or not addressed from prior feedback. Still want to dive a little deeper but wanted to give a first pass. I resolved all issues you fixed, but there are still some remaining from the last review.

jpbland1 commented on src/wh_server_crypto.c:
```diff
@@ -0,0 +1,337 @@
+#include <stdlib.h>  /* For NULL */
+#include <string.h>  /* For memset, memcpy */
+
+#include <arpa/inet.h>
```
I think it's the she functions that need HTONS, I think bill had a note about this but I'll remove for now

jpbland1 commented on src/wh_server_crypto.c:
```diff
@@ -0,0 +1,337 @@
+    return ret;
+}
+
+int _wh_Server_HandleCryptoRequest(whServerContext* server,
+    uint16_t* action, uint8_t* data, uint16_t* size)
```
Action doesn't tell you what error occurred, it tells the cryptocb layer that the action it wanted wasn't finished and then packet->error gives the specific error code

billphipps replied:
The intent of the Group/Action was to identify the kind of request/response, with the expectation that a request of a specific group+action will result in a single response of that same group+action, even under error cases.  There's a check in comm_client that ensures that the response group+action matches the last request group+action.  The action should NOT be changed on the response under normal and error circumstances.  I haven't completed the async cancel handling yet that would allow an additional request or an unexpected response to be sent and handled correctly.  These would use fixed group+action values to make processing simple.

jpbland1 commented on test/wh_test_crypto.c:
```diff
@@ -189,36 +194,35 @@ static void _whClientServerThreadTest(whClientConfig* c_conf,
-            .comm = cs_conf,
-            /*.nvm = NULL, */
+    whServerConfig                  s_conf[1] = {{
+       .comm = cs_conf,
```
renamed to mem

jpbland1 commented on test/wh_test_crypto.c:
```diff
@@ -86,10 +89,12 @@ static void* _whClientTask(void *cf)
         printf("Failed to wc_curve25519_make_key %d\n", ret);
         goto exit;
     }
+    printf("pre 2 wc_curve25519_make_key %d\n", ret);
```
done

jpbland1 commented on test/wh_test.c:
```diff
@@ -18,8 +18,8 @@ int whTest_Unit(void)
     WH_TEST_ASSERT(0 == whTest_Crypto());
-    WH_TEST_ASSERT(0 == whTest_Flash_RamSim());
-    WH_TEST_ASSERT(0 == whTest_NvmFlash());
+    //WH_TEST_ASSERT(0 == whTest_Flash_RamSim());
+    //WH_TEST_ASSERT(0 == whTest_NvmFlash());
```
done

bigbrett commented:
@jpbland1 gonna give this a shot on the tricore

billphipps commented on src/wh_cryptocb.c:
consider changing the name of this to wh_client_cryptocb.c as this is the cryptocb for the HSM client.

jpbland1 replied:
done

billphipps commented on src/wh_server_crypto.c:
```diff
@@ -0,0 +1,339 @@
+        ret = hal_flash_read(addr + WOLFHSM_HEADER_SIZE + keyOffset, out,
+            ctx->nvmMetaCache[i].len);
+    }
+    /* calculate key digest */
```
Recommend to make integrity/auth of keys to be optional and not required at this point.  No changes for now.

billphipps commented on src/wh_server_she.c:
```diff
@@ -0,0 +1,907 @@
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_packet.h"
+
+#if 0
```
Maybe hold off on including this file until we are ready to use it?

billphipps commented on test/user_settings.h:
Still not excited about having a user_settings file here when we could put these defines into the makefile instead.  Leave for now.

billphipps reviewed (dismissed):
Still a number of bug/integration changes will be necessary, but we want to get this pushed in.  Please keep track of the requests for name changes, moving of functions to separate files, and the dependence on POSIX features.  I think many of the small things we can pick up quickly in future PR's, so lets get this in.

jpbland1 commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,159 @@
+                /* read out */
+                else {
+                    info->pk.curve25519kg.key->devCtx =
+                        (void*)0 + packet->pkCurve25519kgRes.keyId;
```
typecasting between two different sized types throws an error so I did this instead, super ugly but everything else I tried threw an error

billphipps replied:
intptr_t is your friend.  :)

jpbland1 commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,159 @@
+#endif
+    int ret = CRYPTOCB_UNAVAILABLE;
+    whClientContext* ctx = inCtx;
+    uint8_t rawPacket[WH_COMM_MTU];
```
done

jpbland1 commented on src/wh_server.c:
```diff
@@ -124,29 +110,105 @@ static int _wh_Server_HandleKeyRequest(whServerContext* server,
-    int rc = 0;
-    switch (action) {
-    /* TODO: Add crypto message handling here */
+#if 0
```
done

billphipps commented:
Here's my list of things that should be addressed in the future crypto increments:
1. Update licensing and copyright to GPLv3 and 2024.  This should be handled as a single global PR, once we have our filenames/structures finalized.
2. remove C++ guards.  This is C99 and we use things like struct x = {0}; for compiler-enforced zeroing of structs, especially on the stack, so we can avoid XMEMSET(0)'s
3. Update filenames and type names to follow the typical convention. 
4. Resolve ntoh() and byteswapping in a utility header.  Provide a config to support use of arpa/inet or byte swap, otherwise provide dumb, unoptimized versions.
5. make sure to validate reasonableness of request/response size/offset values before using them.
6. Eliminate posix dependencies lie sleep()/usleep and replace with XSLEEP XUSLEEP in a utility header.
7. Try to eliminate user_settings.h in the test folder and put defines into the Makefile just for feature enablement
8. RAM usage of the key cache is a problem.  Active memory usage in the server for keys should be the usable keys, not the unparsed DER's
9. Use wolfHSM error codes everywhere except cryptocb implementations.  BADARGS should only be used when local function arguments are not right.  Use ABORTED if a problem is detected after the function is started.
10. add optional static_assert on packet sizes to ensure they are less (including reasonable payloads) than the DATA_LEN
11. Add support for large or purely DMA-based crypto calls.  These should be separately configured so it is clear to the user when DMA is used and when memcpy to a buffer is used. 
12.

jpbland1 commented:
I cant' reply to it but I've switched the action to be unchanged and added an rc field on the packet that explicitly gives the client the return code every time

billphipps commented on wolfhsm/wh_packet.h:
```diff
@@ -0,0 +1,484 @@
+        wh_Packet_she_extend_seed_res sheExtendSeedRes;
+#endif
+    };
+} whPacket;
```
added to todo list

billphipps commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,531 @@
+                do {
+                    ret = wh_Client_RecvResponse(ctx, &group, &action, &dataSz,
+                        rawPacket);
+                    sleep(1);
```
added to todo list

billphipps commented on src/wh_cryptocb.c:
```diff
@@ -0,0 +1,159 @@
+        /* out is after the fixed size fields */
+        out = (uint8_t*)(&packet->rngRes + 1);
+        /* set sz */
+        packet->rngReq.sz = info->rng.sz;
```
Added to todo list

billphipps reviewed (approved):
Make sure to start chewing on the todo list.
